File: Exemplos de Código/Grammar.hs
{-
  Compute NULLABLE, FIRST and FOLLOW mappings from
  a Context-free Grammar

  Pedro Vasconcelos, 2021, 2022
-}
module Grammar where

import           Data.List
import           Data.Char
import qualified Data.Map as Map
import qualified Data.Set as Set
import           Data.Set (Set)
import           Data.Map (Map, (!))

-- Both terminal and non-terminal symbols are represented by strings
type Sym = String

-- A rule is a pair of (lhs, rhs)
-- lhs is a (non-terminal) symbol;
-- rhs is a list of symbols (possibly empty)
type Rule = (Sym, [Sym])

-- A grammar is a list of rules;
-- the first rule should be for the initial symbol
type Grammar = [Rule]

-- Example grammar
example1 = [ ("S'", ["S", "$"])
            , ("S", ["A", "B"])
            , ("A", ["a", "A", "b"])
            , ("A", [])
            , ("B", ["b", "B"])
            , ("B", [])
            ]

-- Grammar for expressions
example2 = [ ("Expr", ["Term", "TermList"])
          , ("TermList", ["+", "Term", "TermList"])
          , ("TermList", [])
          , ("Term", ["Factor", "FactorList"])
          , ("FactorList", ["*", "Factor", "FactorList"])
          , ("FactorList", [])
          , ("Factor", ["num"])
          , ("Factor", ["(", "Expr", ")"])
          ]


type Nullable = Map Sym Bool
type First = Map Sym (Set Sym)
type Follow = Map Sym (Set Sym)


nonTerminals :: Grammar -> [Sym]
nonTerminals = nub . map fst

isNonTerminal :: Sym -> Bool
isNonTerminal (x:_) = isUpper x

-- check if a sequence of symbols is nullable
-- given the nullable mapping for non-terminals
nullableSyms :: Nullable -> [Sym] -> Bool
nullableSyms nullable xs
  = and [isNonTerminal x && nullable!x | x<-xs]


-- compute the FIRST terminals of a sequence of symbols
-- given nullable and first mappings for non-terminals
firstSyms :: Nullable -> First -> [Sym] -> Set Sym
firstSyms nullable first xs = go xs
  where go [] = Set.empty
        go (x:xs) 
          | isNonTerminal x = (first!x) `Set.union`
                              if nullable!x then go xs else Set.empty
          | otherwise = Set.singleton x


-- compute NULLABLE mapping
findNullable :: Grammar -> Nullable
findNullable rules = fixpoint next initial
  where
    vars = nonTerminals rules
    initial = Map.fromList $ zip vars (repeat False)
    next iter = Map.fromListWith
                (||)
                [ (var, nullableSyms iter rhs) | (var,rhs) <- rules ]


-- compute FIRST mapping
findFirst :: Grammar  -> First
findFirst rules = fixpoint next initial
  where
    nullable = findNullable rules
    vars = nonTerminals rules
    initial = Map.fromList $ zip vars (repeat Set.empty)
    next iter = Map.fromListWith Set.union
                [ (var, firstSyms nullable iter rhs) | (var, rhs)<-rules]



-- compute FOLLOW mapping
findFollow :: Grammar -> Follow
findFollow rules = fixpoint next initial 
  where
    vars = nonTerminals rules
    nullable = findNullable rules
    first = findFirst rules
    initial = Map.fromList $ zip vars (repeat Set.empty)
    next iter = Map.fromListWith Set.union
                [ (x,s) | (var, rhs) <- rules
                        , (x:rest) <- tails rhs
                        , isNonTerminal x
                        , s <- [firstSyms nullable first rest] ++
                               [iter!var | nullableSyms nullable rest]
                        ] 
               

-- compute iterations until a fixed point is reached
fixpoint :: Eq a => (a -> a) -> a -> a
fixpoint transf initial = go initial
  where
    go x = let x' = transf x
           in if x==x' then x
              else go x'

File: Code Examples/grammar.py
# 
# Grammar.py  - Computer NULLABLE, FIRST and FOLLOW sets
#
# Pedro Vasconcelos, 2021, 2022

import copy
from typing import List, Dict, Set, Tuple

#
# Context Free Gramar are represented by lists of Rules
# Each rule is a pair of non-terminal (string) and list of symbols (strings)
#
Rule = Tuple[str, List[str]]

# Example grammar
#
example1 = [ ("S'", ["S", "$"])
          , ("S", ["A", "B"])
          , ("A", ["a", "A", "b"])
            , ("A", [])
            , ("B", ["b", "B"])
            , ("B", [])
            ]
            

# Grammar for expressions
#
example2 = [ ("Expr", ["Term", "TermList"])
             , ("TermList", ["+", "Term", "TermList"])
             , ("TermList", [])
             , ("Term", ["Factor", "FactorList"])
             , ("FactorList", ["*", "Factor", "FactorList"])
             , ("FactorList", [])
             , ("Factor", ["num"])
             , ("Factor", ["(", "Expr", ")"])
          ]

#
# NULLABLE, FIRST and FOLLOW are dictionaries
#
Nullable = Dict[str,bool]
First = Dict[str,Set[str]]
Follow = Dict[str,Set[str]]

# auxiliary definitions

def is_nonterminal(sym: str) -> bool:
    return sym[0].isupper()

def is_terminal(sym: str) -> bool:
    return not is_nonterminal(sym)

def non_terminals(rules: List[Rule]) -> Set[str]:
    return { var for (var,rhs) in rules }

# check if a sequence of symbols is nullable
# given the Nullable mapping for non-terminals
def nullable_syms(nullable: Nullable, xs: List[str]) -> bool:
    for x in xs:
        if is_terminal(x) or not nullable[x]:
            return False
    return True

# find the first terminal of a sequence
# given mappings for Nullable and First
def first_syms(nullable: Nullable, first: First, xs: List[str]) -> Set[str]:
    result = set()
    for x in xs:
        if is_nonterminal(x):
            result.update(first[x])
            if not nullable[x]:
                break
        else:
            result.add(x)
            break
    return result

#
# compute NULLABLE mapping
#
def find_nullable(rules: List[Rule]) -> Nullable:
    vars = non_terminals(rules)
    iter = { var:False for var in vars }
    prev = None
    while prev != iter:
        prev = iter.copy()  #  shallow copy can be used here
        for var,rhs in rules:
            iter[var] = iter[var] or nullable_syms(iter, rhs)
    return iter

#
# compute FIRST mapping
#
def find_first(rules: List[Rule]) -> First:
    nullable = find_nullable(rules)
    vars = non_terminals(rules)
    iter : First = { var:set() for var in vars }
    prev = None
    while prev != iter:
        prev = copy.deepcopy(iter) # because of dictionary values are sets
        for var,rhs in rules:
            iter[var].update(first_syms(nullable,iter,rhs))
    return iter

#
# compute FOLLLOW mapping
#
def find_follow(rules: List[Rule]) -> Follow:
    nullable = find_nullable(rules)
    first = find_first(rules)
    vars = non_terminals(rules)
    iter : Follow = { var:set() for var in vars }
    prev = None
    while prev != iter:
        prev = copy.deepcopy(iter) # because dictionary values are sets
        for var,rhs in rules:
            for i in range(len(rhs)):
                x = rhs[i]
                if is_nonterminal(x):
                    rest = rhs[i+1:]
                    iter[x].update(first_syms(nullable,first,rest))
                    if nullable_syms(nullable,rest):
                        iter[x].update(iter[var])
    return iter




